Core/FileSystems/AndroidStorageFileSystem.cpp

```diff
@@ -0,0 +1,539 @@
+	EntryMap::iterator iter = entries.find(handle);
+	if (iter != entries.end()) {
+		return iter->second.hFile.Seek(position, type);
+	} else {
+		//This shouldn't happen...
+		ERROR_LOG(FILESYS, "Cannot seek in file that hasn't been opened: %08x", handle);
+		return 0;
+	}
+}
+
+PSPFileInfo AndroidStorageFileSystem::GetFileInfo(std::string filename) {
+	PSPFileInfo x;
+	x.name = filename;
+
+	std::string uri = GetLocalPath(filename);
+
+	FileInfo info;
+	if (!Android_GetFileInfo(uri, &info)) {
+		return ReplayApplyDiskFileInfo(x, CoreTiming::GetGlobalTimeUs());
+	}
+
+	x.type = info.isDirectory ? FILETYPE_DIRECTORY : FILETYPE_NORMAL;
+	x.exists = true;
+
+	if (x.type != FILETYPE_DIRECTORY) {
+		x.size = info.size;
+		x.access = info.isWritable ? 0777 : 0666;
+
+		// The only time value we get from the storage API.
+		int64_t lastModified = info.lastModified / 1000;
+
+		time_t atime = lastModified;
+		time_t ctime = lastModified;
+		time_t mtime = lastModified;
+
+		localtime_r((time_t*)&atime, &x.atime);
+		localtime_r((time_t*)&ctime, &x.ctime);
+		localtime_r((time_t*)&mtime, &x.mtime);
+	}
+
+	return ReplayApplyDiskFileInfo(x, CoreTiming::GetGlobalTimeUs());
+}
+
+bool AndroidStorageFileSystem::GetHostPath(const std::string &inpath, std::string &outpath) {
+	outpath = GetLocalPath(inpath);
+	return true;
+}
+
+// See comment in DirectoryFileSystem.
+extern std::string SimulateVFATBug(std::string filename);
+
+std::vector<PSPFileInfo> AndroidStorageFileSystem::GetDirListing(std::string path) {
+	std::vector<PSPFileInfo> myVector;
+	bool listingRoot = path == "/" || path == "\\";
+
+	std::string uri = GetLocalPath(path);
+
+	std::vector<FileInfo> fileInfo = Android_ListContentUri(uri);
+
+	bool hideISOFiles = PSP_CoreParameter().compat.flags().HideISOFiles;
+	for (auto &info : fileInfo) {
+		PSPFileInfo entry;
+		if (info.isDirectory)
+			entry.type = FILETYPE_DIRECTORY;
+		else
+			entry.type = FILETYPE_NORMAL;
+		entry.access = info.isWritable ? 0777 : 0666;
+		entry.name = info.name;
+		if (Flags() & FileSystemFlags::SIMULATE_FAT32)
+			entry.name = SimulateVFATBug(entry.name);
+		entry.size = info.size;
+
+		bool hideFile = false;
+		if (hideISOFiles && (endsWithNoCase(entry.name, ".cso") || endsWithNoCase(entry.name, ".iso"))) {
+			// Workaround for DJ Max Portable, see compat.ini.
+			hideFile = true;
+		}
+		int64_t lastModified = info.lastModified / 1000;
+
+		time_t atime = lastModified;
+		time_t ctime = lastModified;
+		time_t mtime = lastModified;
+
+		localtime_r((time_t*)&s.st_atime, &entry.atime);
+		localtime_r((time_t*)&s.st_ctime, &entry.ctime);
+		localtime_r((time_t*)&s.st_mtime, &entry.mtime);
+		if (!hideFile && (!listingRoot || (strcmp(info.name.c_str(), "..") && strcmp(info.name.c_str(), "."))))
+			myVector.push_back(entry);
+	}
+
+	return ReplayApplyDiskListing(myVector, CoreTiming::GetGlobalTimeUs());
+}
+
+u64 AndroidStorageFileSystem::FreeSpace(const std::string &path) {
+	// Can't get this, I think.
```

Comment:
That's gonna result in confusion when people's saves just silently don't work, especially if it's on external storage without alerts of free space.  Sigh.

-[Unknown]